esp/esp/program/modules/handlers/programprintables.py

```diff
@@ -310,6 +311,22 @@ def coursecatalog(self, request, tl, one, two, module, extra, prog):
 
         return render_to_latex(self.baseDir()+template_name, context, extra)
 
+    @aux_call
+    @needs_admin
+    def classpopularity(self, request, tl, one, two, module, extra, prog):
+        classes = ClassSubject.objects.filter(parent_program = prog)
+        priorities = range(1, prog.studentclassregmoduleinfo.priority_limit + 1)
+        for prioritiy in priorities:
+            classes = classes.annotate(**{'priority' + str(prioritiy): Count(
+            Case(When(sections__studentregistration__relationship__name='Priority/' + str(prioritiy), then=1), default=None, output_field=IntegerField()
+            ))})
+        classes = classes.annotate(ssi=Count('studentsubjectinterest', distinct=True))
```

Comment:
Ah, I just learned about a lot of this! Most of it is described in the [django docs](https://docs.djangoproject.com/en/1.8/ref/models/conditional-expressions/), but I can outline what's going on. We have a list of the classes for the program from line 317. Ultimately, we want to know how many students expressed interest in each class and how many students set the class as one of their priorities. In line 323, you can see the simpler version of one of these annotations. We can use the `annotate()` function to basically add a new model field to each object in our list. In this case, we'll call the field `ssi` and we'll count up (using `Count()`) how many `studentsubjectinterest` objects are associated with each class (django knows to do this at a class level, so you don't need to worry about that). I included `distinct=True` in that case because the former lines seemed to cause some weird duplication of the database, inflating the ssi numbers.

Now, for the priority counts it's a little harder. First, we don't know a priori how many priorities students can set, so we can't hard code that. We also ultimately want to loop over the priorities, so we want something that can be based on the looping through a sequence of numbers.

First, we get the number of priorities that students are allowed to pick on line 318. Again, we'll use `annotate()` to  count how many students set this class as a priority in the lottery, but we need to filter the count by the specific priority number. In this case, we also want to save each count as a different field, "priorityX". By default, you can't use strings to assign variable/field names, so we have to use the funky ** notation (discussed [here](https://yuji.wordpress.com/2009/09/12/django-python-dynamically-create-queries-from-a-string-and-the-or-operator/)). Here we set up a dictionary and the key:value pair is then used as the field:data pair in the `annotate()` call. Then, in order to only count the `studentregistration` objects for that specific priority, we use Case(When()). Any time it finds a `studentregistration` that has that priority, it adds it to a list as a "1" (we don't need any actual information about the sr).  Any other time, it adds `None` to the list. Then it essentially takes the length of that list (the `None`s don't count), giving us how many `studentregistration` objects were associated with the class and had that specific priority value. We do that process for each priority level, saving them all as separate model fields for each class. Then we can access those fields when we send the class list to the template.

Reply:
Also, because the fields have different names based on the priority value, that's why I needed to make a new template tag to allow us to access a field based on a string (because `cls."priority" + "1"` doesn't work, and neither does `cls.prior_field` where we set up prior_field to be `"priority" + "1" beforehand).

Reply:
Thank you!